From db5fa3cc089579e5817739660bd4276d07fd288a Mon Sep 17 00:00:00 2001
From: Javier Maestro <jjmaestro@ieee.org>
Date: Sat, 4 Apr 2026 00:34:43 +0000
Subject: [PATCH 07/14] refactor: extract shared download helpers into
 lib/download.bzl

Download logic (integrity checking, download arg building, patch
application) was tangled into the repository rule. Extract it into
lib/download.bzl to separate concerns and make it independently
testable.
---
 download/archives/BUILD.bazel    |   1 +
 download/archives/repository.bzl | 100 +-----------
 lib/BUILD.bazel                  |   9 ++
 lib/download.bzl                 | 108 +++++++++++++
 tests/lib/BUILD.bazel            |   3 +
 tests/lib/download.bzl           | 258 +++++++++++++++++++++++++++++++
 6 files changed, 383 insertions(+), 96 deletions(-)
 create mode 100644 lib/download.bzl
 create mode 100644 tests/lib/download.bzl

diff --git a/download/archives/BUILD.bazel b/download/archives/BUILD.bazel
index 68f343f..ada178f 100644
--- a/download/archives/BUILD.bazel
+++ b/download/archives/BUILD.bazel
@@ -10,6 +10,7 @@ bzl_library(
     name = "repository",
     srcs = ["repository.bzl"],
     deps = [
+        "//lib:download",
         "//lib:index",
         "//lib:utils",
         "@version_utils//spec",
diff --git a/download/archives/repository.bzl b/download/archives/repository.bzl
index f36e1a7..5651a47 100644
--- a/download/archives/repository.bzl
+++ b/download/archives/repository.bzl
@@ -9,6 +9,7 @@ to download and extract an archive: the archive's `version`(s) and the
 
 load("@version_utils//spec:spec.bzl", Spec = "spec")
 load("@version_utils//version:version.bzl", Version = "version")
+load("//lib:download.bzl", Download = "download")
 load("//lib:index.bzl", Index = "index")
 load("//lib:utils.bzl", Utils = "utils")
 
@@ -490,90 +491,6 @@ alias(
 )
 '''
 
-BUILD_PATCHES = '''\
-"""Generated by download_archives. DO NOT EDIT."""
-
-package(default_visibility = ["//visibility:public"])
-
-exports_files(glob(
-    [
-        "*.patch",
-        "*.diff",
-    ],
-    allow_empty = True,
-))
-'''
-
-def _apply_patch(rctx, version, source, patch):
-    patch_content = rctx.read(patch)
-
-    # NOTE:
-    # The patches are generated from the root of the repo but the archives are
-    # extracted into /version/source so the patch has to be modified to match
-    # the new path in the Bazel repo
-    for letter in ("a", "b"):
-        for prefix in ("git %s/", "--- %s/", "+++ %s/"):
-            patch_content = patch_content.replace(
-                prefix % letter,
-                (prefix % letter) + "%s/%s/" % (version, source),
-            )
-
-    patch_path = "patches/%s/%s" % (version, source)
-    rctx.file("%s/BUILD.bazel" % patch_path, BUILD_PATCHES, executable = False)
-
-    patch_file = "%s/%s" % (patch_path, patch.name)
-    rctx.file(patch_file, patch_content, executable = False)
-
-    rctx.patch(patch_file, strip = 1)
-
-def _make_download_args(repo_path, repo_context, _print = print):
-    # https://bazel.build/rules/lib/builtins/repository_ctx#download_and_extract
-    allowed = [
-        "url",
-        "sha256",
-        "type",
-        "strip_prefix",
-        "canonical_id",
-        "auth",
-        "headers",
-        "integrity",
-        "rename_files",
-    ]
-
-    # output and allow_fail cannot be exposed to the users: output is fixed and
-    # allow_fail has to be True to attempt the downloads from the different
-    # sources
-    ignored = ["output", "allow_fail"]
-
-    args = {}
-
-    for arg in allowed + ignored:
-        if arg not in repo_context:
-            continue
-
-        if arg in ignored:
-            msg = "{}: ignoring '{}' in context"
-            _print(msg.format(repo_path, arg))
-            continue
-
-        if (
-            arg == "strip_prefix" and
-            native.bazel_version and native.bazel_version < "8.0.0"
-        ):
-            # https://github.com/bazelbuild/bazel/issues/24034
-            args["stripPrefix"] = repo_context[arg]
-        else:
-            args[arg] = repo_context[arg]
-
-    # Bazel download expects either 'sha256' or 'integrity' but not both.
-    # And, even when we've validated the index to prevent the users from using
-    # both, we are actually adding both to the context, so we need to remove it
-    # here to have compatible arguments for the download method
-    if "sha256" in args and "integrity" in args:
-        args.pop("sha256")
-
-    return args
-
 def _apparent_repo_name(rctx):
     # HACK to get the apparent repo name ("short name"). This should be
     # equivalent to native.package_relative_label but that's only available
@@ -612,16 +529,7 @@ def _impl(rctx, _print = print, _fail = fail):
 
     for version, repos in index.repos.items():
         if rctx.attr.strict_integrity:
-            integrities = {
-                repo.context["integrity"]: None
-                for repo in repos
-                if repo.context.get("integrity", None) != None
-            }
-
-            if len(integrities) >= 2:
-                msg = "Version {} has {} different integrities "
-                msg += "while strict_integrity=True"
-                _fail(msg.format(version, len(integrities)))
+            Download.check_strict_integrity(repos, version, _fail)
 
         for repo in repos:
             if version in repo_lock:
@@ -644,7 +552,7 @@ def _impl(rctx, _print = print, _fail = fail):
                 continue
 
             repo_path = "/".join([repo.version, repo.source])
-            download_args = _make_download_args(repo_path, repo.context)
+            download_args = Download.make_download_args(repo_path, repo.context)
 
             download = rctx.download_and_extract(
                 output = repo_path,
@@ -693,7 +601,7 @@ def _impl(rctx, _print = print, _fail = fail):
             )
 
             for patch in matching:
-                _apply_patch(rctx, repo.version, repo.source, patch)
+                Download.apply_patch(rctx, repo.version, repo.source, patch)
 
             if matching:
                 repo_lock[repo.version]["patches"] = [str(p) for p in matching]
diff --git a/lib/BUILD.bazel b/lib/BUILD.bazel
index 1e27b4c..44eb9b4 100644
--- a/lib/BUILD.bazel
+++ b/lib/BUILD.bazel
@@ -3,12 +3,21 @@ load("@bazel_skylib//:bzl_library.bzl", "bzl_library")
 package(default_visibility = ["//visibility:public"])
 
 exports_files([
+    "download.bzl",
     "index.bzl",
     "integrity.bzl",
     "schema.bzl",
     "utils.bzl",
 ])
 
+bzl_library(
+    name = "download",
+    srcs = ["download.bzl"],
+    deps = [
+        ":utils",
+    ],
+)
+
 bzl_library(
     name = "utils",
     srcs = ["utils.bzl"],
diff --git a/lib/download.bzl b/lib/download.bzl
new file mode 100644
index 0000000..3082296
--- /dev/null
+++ b/lib/download.bzl
@@ -0,0 +1,108 @@
+"""Shared download helpers for archive repository rules."""
+
+_BUILD_PATCHES = '''\
+"""Generated by download_archives. DO NOT EDIT."""
+
+package(default_visibility = ["//visibility:public"])
+
+exports_files(glob(
+    [
+        "*.patch",
+        "*.diff",
+    ],
+    allow_empty = True,
+))
+'''
+
+def _check_strict_integrity(repos, version, _fail):
+    """Fail if repos for a version have different integrity hashes."""
+    integrities = {
+        repo.context["integrity"]: None
+        for repo in repos
+        if repo.context.get("integrity", None) != None
+    }
+
+    if len(integrities) >= 2:
+        msg = "Version {} has {} different integrities "
+        msg += "while strict_integrity=True"
+        return _fail(msg.format(version, len(integrities)))
+
+    return None
+
+def _make_download_args(repo_path, repo_context, _print = print):
+    """Build rctx.download_and_extract kwargs from a materialized repo context."""
+
+    # https://bazel.build/rules/lib/builtins/repository_ctx#download_and_extract
+    allowed = [
+        "url",
+        "sha256",
+        "type",
+        "strip_prefix",
+        "canonical_id",
+        "auth",
+        "headers",
+        "integrity",
+        "rename_files",
+    ]
+
+    # output and allow_fail cannot be exposed to the users: output is fixed and
+    # allow_fail has to be True to attempt the downloads from the different
+    # sources
+    ignored = ["output", "allow_fail"]
+
+    args = {}
+
+    for arg in allowed + ignored:
+        if arg not in repo_context:
+            continue
+
+        if arg in ignored:
+            msg = "{}: ignoring '{}' in context"
+            _print(msg.format(repo_path, arg))
+            continue
+
+        if (
+            arg == "strip_prefix" and
+            native.bazel_version and native.bazel_version < "8.0.0"
+        ):
+            # https://github.com/bazelbuild/bazel/issues/24034
+            args["stripPrefix"] = repo_context[arg]
+        else:
+            args[arg] = repo_context[arg]
+
+    # Bazel download expects either 'sha256' or 'integrity' but not both. And,
+    # even when we've validated the index to prevent the users from using both,
+    # we are actually adding both to the context, so we need to remove it here
+    # to have compatible arguments for the download method
+    if "sha256" in args and "integrity" in args:
+        args.pop("sha256")
+
+    return args
+
+def _apply_patch(rctx, version, source, patch):
+    """Apply a patch with path rewriting for nested extraction."""
+    patch_content = rctx.read(patch)
+
+    # The patches are generated from the root of the repo but the archives are
+    # extracted into /version/source so the patch has to be modified to match
+    # the new path in the Bazel repo
+    for letter in ("a", "b"):
+        for prefix in ("git %s/", "--- %s/", "+++ %s/"):
+            patch_content = patch_content.replace(
+                prefix % letter,
+                (prefix % letter) + "%s/%s/" % (version, source),
+            )
+
+    patch_path = "patches/%s/%s" % (version, source)
+    rctx.file("%s/BUILD.bazel" % patch_path, _BUILD_PATCHES, executable = False)
+
+    patch_file = "%s/%s" % (patch_path, patch.name)
+    rctx.file(patch_file, patch_content, executable = False)
+
+    rctx.patch(patch_file, strip = 1)
+
+download = struct(
+    check_strict_integrity = _check_strict_integrity,
+    make_download_args = _make_download_args,
+    apply_patch = _apply_patch,
+)
diff --git a/tests/lib/BUILD.bazel b/tests/lib/BUILD.bazel
index 23a37d5..d5603be 100644
--- a/tests/lib/BUILD.bazel
+++ b/tests/lib/BUILD.bazel
@@ -1,8 +1,11 @@
+load(":download.bzl", download_test_suite = "test_suite")
 load(":index.bzl", index_test_suite = "test_suite")
 load(":integrity.bzl", integrity_test_suite = "test_suite")
 load(":schema.bzl", schema_test_suite = "test_suite")
 load(":utils.bzl", utils_test_suite = "test_suite")
 
+download_test_suite()
+
 index_test_suite()
 
 integrity_test_suite()
diff --git a/tests/lib/download.bzl b/tests/lib/download.bzl
new file mode 100644
index 0000000..94a8e8b
--- /dev/null
+++ b/tests/lib/download.bzl
@@ -0,0 +1,258 @@
+"download.bzl unit tests"
+
+load("@bazel_skylib//lib:unittest.bzl", "asserts", "unittest")
+load("//lib:download.bzl", Download = "download")
+load("//tests:mock.bzl", Mock = "mock")
+load("//tests:suite.bzl", _test_suite = "test_suite")
+
+def _mock_fail(errors):
+    """Returns a _fail function that appends errors to a list."""
+    return lambda msg: errors.append(msg)
+
+def _check_strict_integrity__all_same_impl(ctx):
+    env = unittest.begin(ctx)
+
+    repos = [
+        struct(context = {"integrity": "sha256-AAA"}),
+        struct(context = {"integrity": "sha256-AAA"}),
+    ]
+
+    errors = []
+    Download.check_strict_integrity(repos, "1.0", _fail = _mock_fail(errors))
+    asserts.equals(env, 0, len(errors))
+
+    return unittest.end(env)
+
+check_strict_integrity__all_same_test = unittest.make(
+    _check_strict_integrity__all_same_impl,
+)
+
+def _check_strict_integrity__mismatch_impl(ctx):
+    env = unittest.begin(ctx)
+
+    repos = [
+        struct(context = {"integrity": "sha256-AAA"}),
+        struct(context = {"integrity": "sha256-BBB"}),
+    ]
+
+    errors = []
+    Download.check_strict_integrity(repos, "1.0", _fail = _mock_fail(errors))
+    asserts.equals(env, 1, len(errors))
+    asserts.true(env, "different integrities" in errors[0])
+
+    return unittest.end(env)
+
+check_strict_integrity__mismatch_test = unittest.make(
+    _check_strict_integrity__mismatch_impl,
+)
+
+def _check_strict_integrity__mix_impl(ctx):
+    env = unittest.begin(ctx)
+
+    repos = [
+        struct(context = {"integrity": "sha256-AAA"}),
+        struct(context = {}),
+    ]
+
+    # one has integrity, one doesn't: should not fail (only 1 unique integrity)
+    errors = []
+    Download.check_strict_integrity(repos, "1.0", _fail = _mock_fail(errors))
+    asserts.equals(env, 0, len(errors))
+
+    return unittest.end(env)
+
+check_strict_integrity__mix_test = unittest.make(
+    _check_strict_integrity__mix_impl,
+)
+
+def _check_strict_integrity__none_impl(ctx):
+    env = unittest.begin(ctx)
+
+    repos = [
+        struct(context = {}),
+        struct(context = {}),
+    ]
+
+    # no integrities at all: should not fail
+    errors = []
+    Download.check_strict_integrity(repos, "1.0", _fail = _mock_fail(errors))
+    asserts.equals(env, 0, len(errors))
+
+    return unittest.end(env)
+
+check_strict_integrity__none_test = unittest.make(
+    _check_strict_integrity__none_impl,
+)
+
+def _make_download_args__basic_impl(ctx):
+    env = unittest.begin(ctx)
+
+    context = {"url": "https://example.com/foo.tar.gz"}
+
+    args = Download.make_download_args("1.0/src", context)
+    asserts.equals(env, {"url": "https://example.com/foo.tar.gz"}, args)
+
+    return unittest.end(env)
+
+make_download_args__basic_test = unittest.make(_make_download_args__basic_impl)
+
+def _make_download_args__sha256_integrity_conflict_impl(ctx):
+    env = unittest.begin(ctx)
+
+    context = {
+        "integrity": "sha256-AAAA",
+        "sha256": "deadbeef",
+        "url": "https://example.com/foo.tar.gz",
+    }
+
+    args = Download.make_download_args("1.0/src", context)
+
+    # integrity wins, sha256 is dropped
+    asserts.true(env, "integrity" in args)
+    asserts.false(env, "sha256" in args)
+    asserts.equals(env, "sha256-AAAA", args["integrity"])
+
+    return unittest.end(env)
+
+make_download_args__sha256_integrity_conflict_test = unittest.make(
+    _make_download_args__sha256_integrity_conflict_impl,
+)
+
+def _make_download_args__ignored_keys_impl(ctx):
+    env = unittest.begin(ctx)
+
+    logs = []
+
+    context = {
+        "allow_fail": True,
+        "output": "/some/path",
+        "url": "https://example.com/foo.tar.gz",
+    }
+
+    args = Download.make_download_args("1.0/src", context, _print = Mock.print(logs))
+
+    # output and allow_fail should be ignored (not in result)
+    asserts.false(env, "output" in args)
+    asserts.false(env, "allow_fail" in args)
+    asserts.equals(env, {"url": "https://example.com/foo.tar.gz"}, args)
+
+    # warnings should be printed
+    asserts.equals(env, 2, len(logs))
+
+    return unittest.end(env)
+
+make_download_args__ignored_keys_test = unittest.make(
+    _make_download_args__ignored_keys_impl,
+)
+
+def _make_download_args__unknown_keys_skipped_impl(ctx):
+    env = unittest.begin(ctx)
+
+    context = {
+        "repo_name": "my_repo",
+        "source": "github.com",
+        "url": "https://example.com/foo.tar.gz",
+        "version": "1.0",
+    }
+
+    args = Download.make_download_args("1.0/src", context)
+
+    # only url should pass through; repo_name, version, source are not
+    # in the allowed list
+    asserts.equals(env, {"url": "https://example.com/foo.tar.gz"}, args)
+
+    return unittest.end(env)
+
+make_download_args__unknown_keys_skipped_test = unittest.make(
+    _make_download_args__unknown_keys_skipped_impl,
+)
+
+def _make_download_args__allowed_keys_passthrough_impl(ctx):
+    env = unittest.begin(ctx)
+
+    # NOTE: strip_prefix is excluded because _make_download_args checks
+    # native.bazel_version for the Bazel 7 compat rename, and
+    # native.bazel_version is not available in test (BUILD) context.
+    context = {
+        "canonical_id": "foo-1.0",
+        "integrity": "sha256-AAAA",
+        "type": "tar.gz",
+        "url": "https://example.com/foo.tar.gz",
+    }
+
+    args = Download.make_download_args("1.0/src", context)
+    asserts.equals(env, "https://example.com/foo.tar.gz", args["url"])
+    asserts.equals(env, "sha256-AAAA", args["integrity"])
+    asserts.equals(env, "tar.gz", args["type"])
+    asserts.equals(env, "foo-1.0", args["canonical_id"])
+
+    return unittest.end(env)
+
+make_download_args__allowed_keys_passthrough_test = unittest.make(
+    _make_download_args__allowed_keys_passthrough_impl,
+)
+
+# --- apply_patch ---
+
+_SAMPLE_PATCH = """\
+diff --git a/foo.c b/foo.c
+--- a/foo.c
++++ b/foo.c
+@@ -1 +1 @@
+-old
++new
+"""
+
+def _apply_patch__rewrites_paths_impl(ctx):
+    env = unittest.begin(ctx)
+
+    files = {}
+    patches = []
+
+    def _read(_label):
+        return _SAMPLE_PATCH
+
+    def _file(path, content = "", **_kwargs):
+        files[path] = content
+
+    def _patch(path, **_kwargs):
+        patches.append(path)
+
+    rctx = struct(read = _read, file = _file, patch = _patch)
+    patch_label = struct(name = "fix.patch")
+
+    Download.apply_patch(rctx, "1.0", "github.com", patch_label)
+
+    # The rewritten patch should have version/source/ prepended to paths
+    rewritten = files["patches/1.0/github.com/fix.patch"]
+    asserts.true(env, "--- a/1.0/github.com/foo.c" in rewritten)
+    asserts.true(env, "+++ b/1.0/github.com/foo.c" in rewritten)
+
+    # The original unprefixed paths should NOT be present
+    asserts.false(env, "--- a/foo.c" in rewritten)
+
+    # Patch was applied
+    asserts.equals(env, 1, len(patches))
+
+    return unittest.end(env)
+
+apply_patch__rewrites_paths_test = unittest.make(
+    _apply_patch__rewrites_paths_impl,
+)
+
+TEST_SUITE_NAME = "lib/download"
+
+TEST_SUITE_TESTS = {
+    "apply_patch/rewrites_paths": apply_patch__rewrites_paths_test,
+    "check_strict_integrity/all_same": check_strict_integrity__all_same_test,
+    "check_strict_integrity/mismatch": check_strict_integrity__mismatch_test,
+    "check_strict_integrity/mix": check_strict_integrity__mix_test,
+    "check_strict_integrity/none": check_strict_integrity__none_test,
+    "make_download_args/allowed_keys_passthrough": make_download_args__allowed_keys_passthrough_test,
+    "make_download_args/basic": make_download_args__basic_test,
+    "make_download_args/ignored_keys": make_download_args__ignored_keys_test,
+    "make_download_args/sha256_integrity_conflict": make_download_args__sha256_integrity_conflict_test,
+    "make_download_args/unknown_keys_skipped": make_download_args__unknown_keys_skipped_test,
+}
+
+test_suite = lambda: _test_suite(TEST_SUITE_NAME, TEST_SUITE_TESTS)
